source/particle/property/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,1102 @@
+            Assert(std::isfinite(inv_sum_volume_mineral),
+                   ExcMessage("inv_sum_volume_mineral is not finite. sum_volume_enstatite = "
+                              + std::to_string(sum_volume_mineral)));
+
+            for (unsigned int grain_i = 0; grain_i < n_grains; ++grain_i)
+              {
+                volume_fractions_grains[mineral_i][grain_i] *= inv_sum_volume_mineral;
+                Assert(isfinite(volume_fractions_grains[mineral_i][grain_i]),
+                       ExcMessage("volume_fractions_grains[mineral_i]" + std::to_string(grain_i) + "] is not finite: "
+                                  + std::to_string(volume_fractions_grains[mineral_i][grain_i]) + ", inv_sum_volume_mineral = "
+                                  + std::to_string(inv_sum_volume_mineral) + "."));
+
+
+                for (size_t j = 0; j < 3; j++)
+                  {
+                    for (size_t k = 0; k < 3; k++)
+                      {
+                        Assert(!std::isnan(rotation_matrices_grains[mineral_i][grain_i][j][k]), ExcMessage(" rotation_matrices_grains is nan before orthoganalization."));
+                      }
+
+                  }
+              }
+
+
+            /**
+             * Correct direction cosine matrices numerical error (orthnormality) after integration
+             * Follows same method as in matlab version from Thissen (see https://github.com/cthissen/Drex-MATLAB/)
+             * of finding the nearest orthonormal matrix using the SVD
+             */
+            for (unsigned int grain_i = 0; grain_i < n_grains; ++grain_i)
+              {
+                rotation_matrices_grains[mineral_i][grain_i] = dealii::project_onto_orthogonal_tensors(rotation_matrices_grains[mineral_i][grain_i]);
+                for (size_t i = 0; i < 3; i++)
+                  for (size_t j = 0; j < 3; j++)
+                    {
+                      // I don't think this should happen with the projection, but D-Rex
+                      // does not do the orthogonal projection, but just clamps the values
+                      // to 1 and -1.
+                      Assert(std::fabs(rotation_matrices_grains[mineral_i][grain_i][i][j]) <= 1.0,
+                             ExcMessage("The rotation_matrices_grains[mineral_i] has a entry asolute larger than 1."));
+                    }
+              }
+
+            for (unsigned int grain_i = 0; grain_i < n_grains; ++grain_i)
+              {
+
+                Assert(isfinite(volume_fractions_grains[mineral_i][grain_i]),
+                       ExcMessage("volume_fractions_grains[mineral_i][" + std::to_string(grain_i) + "] is not finite: "
+                                  + std::to_string(volume_fractions_grains[mineral_i][grain_i]) + ", inv_sum_volume_grains[mineral_i] = "
+                                  + std::to_string(inv_sum_volume_mineral) + "."));
+
+                for (size_t j = 0; j < 3; j++)
+                  for (size_t k = 0; k < 3; k++)
```

Comment:
There are several of these loops in this PR, please change the increment operator to a pre-increment (`++j` instead of `j++`). It doesnt matter for these application, just for consistency, because we do it that way everywhere if possible.